src/dawn/engine_dawn.cc

```diff
@@ -179,8 +189,253 @@ MapResult MapBuffer(const ::dawn::Device& device, const ::dawn::Buffer& buf) {
+                             .buffer->GetTexelStride();
+  const auto dawn_row_pitch = Align(width * pixelSize, kMinimumImageRowPitch);
+  const auto fb_buffer_size = dawn_row_pitch * height + width * pixelSize;
+
+  ::dawn::BufferDescriptor descriptor;
+  descriptor.size = fb_buffer_size;
+  descriptor.usage =
+      ::dawn::BufferUsageBit::MapRead | ::dawn::BufferUsageBit::TransferDst;
+  ::dawn::Buffer fb_buffer = device.CreateBuffer(&descriptor);
+
+  ::dawn::TextureCopyView textureCopyView;
+  textureCopyView.texture = render_pipeline.fb_texture;
+  textureCopyView.level = 0;
+  textureCopyView.slice = 0;
+  textureCopyView.origin = {0, 0, 0};
+
+  ::dawn::BufferCopyView bufferCopyView;
+  bufferCopyView.buffer = fb_buffer;
+  bufferCopyView.offset = 0;
+  bufferCopyView.rowPitch = dawn_row_pitch;
+  bufferCopyView.imageHeight = 0;
+
+  ::dawn::Extent3D copySize = {width, height, 1};
+
+  auto encoder = device.CreateCommandEncoder();
+  encoder.CopyTextureToBuffer(&textureCopyView, &bufferCopyView, &copySize);
+
+  auto commands = encoder.Finish();
+  auto queue = device.CreateQueue();
+  queue.Submit(1, &commands);
+
+  MapResult map = MapBuffer(device, fb_buffer);
+  const std::vector<amber::Pipeline::BufferInfo>& out_color_attachment =
+      render_pipeline.pipeline->GetColorAttachments();
+
+  for (size_t i = 0; i < out_color_attachment.size(); ++i) {
+    auto& info = out_color_attachment[i];
+    auto* values = info.buffer->ValuePtr();
+    auto row_stride = pixelSize * width;
+    assert(row_stride * height == info.buffer->GetSizeInBytes());
+    // Each Dawn row has enough data to fill the target row.
+    assert(dawn_row_pitch >= row_stride);
+    values->resize(info.buffer->GetSizeInBytes());
+    // Copy the framebuffer contents back into the host-side
+    // framebuffer-buffer. In the Dawn buffer, the row stride is a multiple of
+    // kMinimumImageRowPitch bytes, so it might have padding therefore memcpy
+    // is done row by row.
+    for (uint h = 0; h < height; h++) {
+      std::memcpy(values->data() + h * row_stride,
+                  static_cast<const uint8_t*>(map.data) + h * dawn_row_pitch,
+                  row_stride);
+    }
+  }
+  // Always unmap the buffer at the end of the engine's command.
+  fb_buffer.Unmap();
+
+  return map;
+}
+
+::dawn::Buffer CreateBufferFromData(const ::dawn::Device& device,
+                                    const void* data,
+                                    uint64_t size,
+                                    ::dawn::BufferUsageBit usage) {
+  ::dawn::BufferDescriptor descriptor;
+  descriptor.size = size;
+  descriptor.usage = usage | ::dawn::BufferUsageBit::TransferDst;
+
+  ::dawn::Buffer buffer = device.CreateBuffer(&descriptor);
+  buffer.SetSubData(0, size, reinterpret_cast<const uint8_t*>(data));
+  return buffer;
+}
+
+::dawn::SamplerDescriptor GetDefaultSamplerDescriptor() {
+  ::dawn::SamplerDescriptor desc;
+
+  float kLodMin = 0.0;
```

Comment:
Move these up to the anonymous namespace at the top of the file if they are constants. (and add a const at the beginning.)